Forbids foo.cdn.ampproject.org.badguys.com from being counted as AMP CDN origin

jonkeller opened the pull request.

jonkeller commented on src/config.js:
```diff
@@ -36,7 +36,7 @@ export const urls = {
   cdn: env['cdnUrl'] || 'https://cdn.ampproject.org',
   cdnProxyRegex: cdnProxyRegex ||
-      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/,
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org$/,
```
This will make the regex no longer match foo.cdn.ampproject.org.badguys.com, and also make it stop matching foo.cdn.ampproject.org/some/legit/file.html
Searched the code for use of this regex, and aside from a stub, all I found was the following, which is testing an origin (e.g. no path/filename) against the regex: https://github.com/ampproject/amphtml/blob/d230337dd356276492d9255c2cc2ce6e4dbce97d/src/url.js#L315
So, at least at this point in time, this shouldn't hurt anything. If in the future we want to test full URLs to see if they come from the CDN, then we should change this to allow an optional slash and path.

jonkeller commented:
Fails amphtml/test/functional/test-cookies.js with:
AssertionError: expected [Function] to throw error matching /Should never attempt to set cookie on proxy origin\: c\&1/ but got 'Should never attempt to set cookie on proxy origin. (in depth check): c&1'

I am assuming it is safe to just change the error message expectation.

jonkeller commented on test/functional/test-cookies.js:
```diff
@@ -109,18 +109,18 @@ describe('cookies', () => {
     expect(() => {
       test('cdn.ampproject.org', 'ampproject.org', true);
-    }).to.throw(/Should never attempt to set cookie on proxy origin\: c\&1/);
+    }).to.throw(/Should never attempt to set cookie on proxy origin.*\: c\&1/);
```
...could have just changed to match line 126 instead, since the full message is:
Should never attempt to set cookie on proxy origin. (in depth check): c&1

Would appreciate close review by someone who knows why lines 112, 115, 118, 123 were originally written differently from 126, 129, 132.

jridgewell replied:
The reason this is happening is because of the protocol used. We need to update the regex to match the `:` part of the origin, too.

jonkeller replied:
(Assuming you meant "port" rather than "protocol".)
Ok, fixed that, to allow the specification of any port, or none.

jonkeller replied:
I also put the unit tests back the way they were.

taymonbeal replied:
Why would we ever need to check against a nonstandard port? Surely we'd never expose one on the AMP Cache?

jonkeller replied:
One would hope not.
Though port 8000 is hard-coded on line 88 of this file. I tested removing ":8000" from that line, and changing the regex in src/config.js back to /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org$/ and all tests in this file pass. I've not pushed that yet, because it makes me curious to know why port 8000 was used in the test in the first place?
@jridgewell is it safe to remove ":8000" from line 88 but then add a test that should fail if given an unusual port?

cramforce commented on src/config.js:
```diff
@@ -36,7 +36,7 @@ export const urls = {
   cdn: env['cdnUrl'] || 'https://cdn.ampproject.org',
   cdnProxyRegex: cdnProxyRegex ||
-      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/,
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org(\:\d+)?$/,
```
I don't think anchoring it in the end is right. Instead we expect either `:` or `/` and then don't care.

taymonbeal replied:
This is checked against origins, not URLs.

jonkeller replied:
Agreed, this is (currently) only ever checked against url.origin, which can not contain the '/' after the hostname, it's just proto://host[:port]
Happy to add a unit test, pending resolution of the comment thread on the file which I changed+reverted.

cramforce replied:
In that case it works for me. Maybe note that as a comment.

jonkeller replied:
Thanks, will do.

jonkeller commented:
@cramforce I believe I have addressed all of your concerns now.
@jridgewell ...but that means I have made changes since your approval. Would you mind taking another look and commenting if you still approve?
Thanks!

cramforce commented on test/functional/test-url.js:
```diff
@@ -429,6 +430,10 @@ describe('isProxyOrigin', () => {
   testProxyOrigin(
       'http://cdn.ampproject.org/', false);
   testProxyOrigin(
+    'https://cdn.ampproject.org.badguys.com/', false);
```
Please add this case `https://cdn.ampproject.orgbadguys.com/`

jonkeller replied:
Done
